Workiva/over_react
# FED-2156 Unsafe required props lint

## Motivation
While required late props are validated to be present when a component is rendered, there are cases where they won't always be present. For example, "partial" props objects that represent a subset of a component's props.

If required late props are accessed without being set, they'll throw at runtime, and there's currently no static analysis that prevents consumers from doing so, or notifies them that they're doing something unsafe.

So, we want to help prevent consumers from unsafely accessing required late props.

## Changes
- Add new `UnsafeRequiredPropAccessDiagnostic` that warns whenever required props are accessed in an unsafe manner
    - Lints when accessing required props on objects that aren't props a component was rendered with
    - Exceptions where we don't lint
        - Accesses on callback args in utility methods like `.getPropKey((p) => p.requiredProp)`
        - Within `memo` areEqual functions, or top-level functions with `areEqual` in the name since consumers apparently split themout a fair bit
        - In files within the `test/` directory, which typically uses `getProps` a lot (and if unsafe accesses happen, they'll just fail tests instead of crashing an app)
        - Within lifecycle methods that receive props as an argument and typically use `typedPropsFactory`
            - We could track these down and ignore them only if they're safe, but that didn't seem worth the effort/complexity
- Add tests
   - Update `SharedAnalysisContext` to allow creation of files outside of `lib`
- Add playground example

#### Release Notes
  

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
        - Verify CI passes
        - Manually test
            - Test in playground example
            - Test in a consumer that uses required props (e.g., FluxUiProps) and verify there aren't any unexpected errors
            - Double-check with `// debug: over_react_metrics` that performance is acceptable (this lint has to process all property accesses to see if they're on props). From my testing it looked pretty efficient, all things considered, and was hovering around being our 5th slowest lint, which IMO isn't bad at all.

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:


[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria
